crypto/http/http_client.c

```diff
@@ -0,0 +1,1200 @@
+
+    switch (retcode) {
+    case HTTP_STATUS_CODE_OK:
+    case HTTP_STATUS_CODE_MOVED_PERMANENTLY:
+    case HTTP_STATUS_CODE_FOUND:
+        return retcode;
+    default:
+        if (retcode < 400)
+            HTTPerr(0, HTTP_R_STATUS_CODE_UNSUPPORTED);
+        else
+            HTTPerr(0, HTTP_R_SERVER_SENT_ERROR);
+        if (*q == '\0')
+            ERR_add_error_data(2, "Code=", p);
+        else
+            ERR_add_error_data(4, "Code=", p, ",Reason=", q);
+        return 0;
+    }
+}
+
+static int check_set_resp_len(OSSL_HTTP_REQ_CTX *rctx, unsigned long len)
+{
+    const char *tag = NULL;
+    long val;
+
+    if (len > rctx->max_resp_len) {
+        HTTPerr(0, HTTP_R_MAX_RESP_LEN_EXCEEDED);
+        tag = ",max=";
+        val = rctx->max_resp_len;
+    }
+    if (rctx->resp_len != 0 && rctx->resp_len != len) {
+        HTTPerr(0, HTTP_R_INCONSISTENT_CONTENT_LENGTH);
+        tag = ",before=";
+        val = rctx->resp_len;
+    }
+    if (tag != NULL) {
+        char len_str[32];
+        char str[32];
+        BIO_snprintf(len_str, sizeof(len_str), "%ld", len);
+        BIO_snprintf(str, sizeof(str), "%ld", val);
+        ERR_add_error_data(4, "length=", len_str, tag, str);
+        return 0;
+    }
+    rctx->resp_len = len;
+    return 1;
+}
+
+/*
+ * Try exchanging request and response via HTTP on (non-)blocking BIO in rctx.
+ * Returns 1 on success, 0 on error or redirection, -1 on BIO_should_retry.
+ */
+int OSSL_HTTP_REQ_CTX_nbio(OSSL_HTTP_REQ_CTX *rctx)
+{
+    int i;
+    long n, n_to_send;
+    unsigned long resp_len;
+    const unsigned char *p;
+    char *key, *value, *line_end;
+
+    if (rctx == NULL) {
+        HTTPerr(0, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    rctx->redirection_url = NULL;
+ next_io:
+    if (!(rctx->state & OHS_NOREAD)) {
+        n = BIO_read(rctx->io, rctx->iobuf, rctx->iobuflen);
+        if (n <= 0) {
+            if (BIO_should_retry(rctx->io))
+                return -1;
+            return 0;
+        }
+
+        /* Write data to memory BIO */
+        if (BIO_write(rctx->mem, rctx->iobuf, n) != n)
+            return 0;
+    }
+
+    switch (rctx->state) {
+    case OHS_HTTP_HEADER:
+        /* Last operation was adding headers: need a final \r\n */
+        if (BIO_write(rctx->mem, "\r\n", 2) != 2) {
+            rctx->state = OHS_ERROR;
+            return 0;
+        }
+        rctx->state = OHS_WRITE_INIT;
+
+        /* fall thru */
+    case OHS_WRITE_INIT:
+        n_to_send = BIO_get_mem_data(rctx->mem, NULL);
+        rctx->state = OHS_WRITE;
+
+        /* fall thru */
+    case OHS_WRITE:
+        n = BIO_get_mem_data(rctx->mem, &p);
+
+        i = BIO_write(rctx->io, p + (n - n_to_send), n_to_send);
+
+        if (i <= 0) {
+            if (BIO_should_retry(rctx->io))
+                return -1;
+            rctx->state = OHS_ERROR;
+            return 0;
+        }
+
+        n_to_send -= i;
+
+        if (n_to_send > 0)
+            goto next_io;
```

Comment:
The various `goto`'s in this state machine, which I took over with various improvements and extensions from `crypto/ocsp/ocsp_ht.c` would be very hard to avoid - I had already tried removing `goto`'s here.